### PR TITLE
Update CI and add badges to README

### PR DESCRIPTION
**Pull request by ChrisHines**
CI updates:

- Update Go versions (1.17.x and 1.18.x)
- Update actions versions
- Use actions for staticcheck
- Drop golint (deprecated)
- Report test coverage

### CODE REVIEW

**Comment by ChrisHines** [comment]
It looks like we'll have to wait for staticcheck to support Go 1.18. It's coming. https://twitter.com/dominikhonnef/status/1503793590991994882

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -1,38 +1,49 @@
...
+  pull_request:
+    types: [synchronize]
 name: Test
 jobs:
   test:
     strategy:
       fail-fast: false
       matrix:
-        go-version: [1.16.x, 1.17.x]
+        go-version: [1.17.x, 1.18.x]
```
Comment (peterbourgon) [inline .github/workflows/test.yml:11]: What do you think about `1.x`?

Comment (ChrisHines) [reply .github/workflows/test.yml:11]: Does that work? What versions would it map to?

Comment (peterbourgon) [reply .github/workflows/test.yml:11]: It works, and maps to the most-recent version, modulo some lag in updating the action. Although, upon reflection, if we commit to supporting the most-recent and next-most-recent version, it doesn't help much there. Maybe not a good idea.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -1,38 +1,49 @@
-on: push
+on:
+  push:
+  pull_request:
+    types: [synchronize]
```
Comment (peterbourgon) [inline .github/workflows/test.yml:4]: Ooh, this is new to me. Where can I learn more?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -1,38 +1,49 @@
...
       - name: Test
         run: go test -race ./...
+      - name: Test coverage
+        run: go test -coverprofile="cover.out" ./... # quotes needed for powershell
+      - name: Send coverage
+        uses: shogo82148/actions-goveralls@v1
+        with:
+          path-to-profile: cover.out
+          flag-name: go${{ matrix.go-version }}-${{ matrix.os }}
+          parallel: true
```
Comment (peterbourgon) [inline .github/workflows/test.yml:41]: I took this out of the main repo, I think, at some point, due to flakiness. Do you have more confidence in it now?

File: .github/workflows/test.yml (hunk 1 above)
Comment (ChrisHines) [reply .github/workflows/test.yml:11]: OK. I vote we keep an explicit version list here then.

File: .github/workflows/test.yml (hunk 3 above)
Comment (ChrisHines) [reply .github/workflows/test.yml:41]: I copied this from `github.com/go-logfmt/logfmt` where it's been working fine for many months. I have that repo set to rerun the actions every 10 days even without any commits and haven't had any issues.

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -1,38 +1,49 @@
...
       - name: Fmt
         if: matrix.platform != 'windows-latest' # :(
         run: "diff <(gofmt -d .) <(printf '')"
         shell: bash
       - name: Vet
         run: go vet ./...
       - name: Staticcheck
-        run: staticcheck ./...
-      - name: Lint
-        run: golint ./...
```
Comment (peterbourgon) [inline .github/workflows/test.yml:36]: Lint seems to just go away. Is that what we want?

Comment (ChrisHines) [reply .github/workflows/test.yml:36]: According to the [golint](https://github.com/golang/lint) readme:

>NOTE: Golint is [deprecated and frozen](https://github.com/golang/go/issues/38968). There's no drop-in replacement for it, but tools such as [Staticcheck](https://staticcheck.io/) and go vet should be used instead.

So, yes, IMO we should stop using it. But Staticcheck's style checks are not usually enabled by default, so we might need to change something if we want its more lint-like features. Something to pass the equivalent of `-checks=all` perhaps. That can also be accomplished by a [staticcheck.conf file](https://staticcheck.io/docs/configuration/#configuration-files) to configure it for anyone running `staticcheck` on a clone of the repo as well.

Comment (peterbourgon) [reply .github/workflows/test.yml:36]: Sure, understood. I'll propose `gofumpt` in a separate PR.

**Comment by ChrisHines** [comment]
Staticcheck has a [new release](https://staticcheck.io/changes/2022.1/) compatible with Go 1.18. I've re-run the checks here and they are passing now. 🎉

File: .github/workflows/test.yml (hunk 4 above)
Comment (ChrisHines) [reply .github/workflows/test.yml:36]: Go for it, I'm a fan of `gofumpt`.

**Comment by ChrisHines** [comment]
This PR should be read to go.

**Review by peterbourgon — APPROVED** [review]

**Pull request merged**